api/src/opentrons/protocol_engine/resources/file_provider.py

```
@@ -0,0 +1,153 @@
+            rows.append(row)
+        for i in range(3):
+            rows.append([""])
+        rows.append(["", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"])
+        for i in range(8):
+            row = [plate_alpharows[i]]
+            for j in range(12):
+                row.append("")
+            rows.append(row)
+        for i in range(3):
+            rows.append([""])
+        rows.append(
+            [
+                "",
+                "ID",
+                "Well",
+                "Absorbance (OD)",
+                "Mean Absorbance (OD)",
+                "Absorbance %CV",
+            ]
+        )
+        for i in range(3):
+            rows.append([""])
+        rows.append(
+            [
+                "",
+                "ID",
+                "Well",
+                "Absorbance (OD)",
+                "Mean Absorbance (OD)",
+                "Dilution Factor",
+                "Absorbance %CV",
+            ]
+        )
+        rows.append(["1", "Sample 1", "", "", "", "1", "", "", "", "", "", ""])
+        for i in range(3):
+            rows.append([""])
+
+        # end of file metadata
+        rows.append(["Protocol"])
+        rows.append(["Assay"])
+        rows.append(["Sample Wavelength (nm)", str(measurement.wavelength)])
+        if self.reference_wavelength is not None:
+            rows.append(["Reference Wavelength (nm)", str(self.reference_wavelength)])
+        rows.append(["Serial No.", self.serial_number])
+        rows.append(["Measurement started at", str(self.start_time)])
+        rows.append(["Measurement finished at", str(self.finish_time)])
+
+        # Ensure the filename adheres to ruleset contains the wavelength for a given measurement
+        filename = _csv_filename_validation(filename)
+        if filename.endswith(".csv"):
+            filename = filename[:-4]
+        filename = filename + "_" + str(measurement.wavelength) + ".csv"
```

A few things about this:

1. Just as a low-level implementation thing, `_csv_filename_validation` is adding a `.csv` suffix that we're immediately stripping here. Can that be simplified to avoid the back-and-forth?
2. Is it the case that if I write `filename="foo.csv"` in my protocol, this will rewrite it as `foo_123.csv`, where `123` is the wavelength? If so, I'd find that pretty confusing. If we must include the wavelength in these filenames, can we have protocol authors write something like `filename_prefix="foo"` instead of `filename="foo.csv"`?

---

Yeah I think that makes sense. Since we are going to generate a file for each of the wavelengths the user scanned the plate with, we have to append the wavelength name to their file name to be in keeping with what users expect to get out of the original device application. Communicating that modification to them in the parameter or parameter description accepting a filename is probably a good idea. I suppose we should just raise an error if they provide a '.' type ending to their filename at all?

---

upon thinking on it, I think we'll reflect that they're providing the first portion of the final filename in the parameter name and parameter description. If they do provide a .csv (which a user may think is a natural thing to do) I won't raise an error, I'll just chop it off and apprend the wavelength data and file ending when we create our final CSV file data set. So long as the description communicates that tot eh user I see no harm in doing this cleanup work for them.